osu.Game/Overlays/Comments/CommentEditor.cs

```diff
@@ -134,18 +142,42 @@ private void load(OverlayColourProvider colourProvider)
             });
 
             TextBox.OnCommit += (_, _) => commitButton.TriggerClick();
+            User.BindTo(API.LocalUser);
         }
 
         protected override void LoadComplete()
         {
             base.LoadComplete();
             Current.BindValueChanged(_ => updateCommitButtonState(), true);
+            User.BindValueChanged(_ => updateStateForLoggedIn(), true);
         }
 
         protected abstract void OnCommit(string text);
 
-        private void updateCommitButtonState() =>
-            commitButton.Enabled.Value = loadingSpinner.State.Value == Visibility.Hidden && !string.IsNullOrEmpty(Current.Value);
+        private void commitOrLogIn(string text)
+        {
+            if (!API.IsLoggedIn)
+            {
+                loginOverlay?.Show();
+                return;
+            }
+
+            OnCommit(text);
+        }
+
+        private void updateCommitButtonState()
+        {
+            bool textBoxValid = loadingSpinner.State.Value == Visibility.Hidden && !string.IsNullOrEmpty(Current.Value);
+            commitButton.Enabled.Value = textBoxValid || !API.IsLoggedIn;
+        }
+
+        private void updateStateForLoggedIn()
+        {
+            TextBox.PlaceholderText = GetPlaceholderText(API.IsLoggedIn);
+            TextBox.ReadOnly = !API.IsLoggedIn;
+            commitButton.Text = GetCommitButtonText(API.IsLoggedIn);
+            updateCommitButtonState();
```

Comment:
Nitpick but it's weird that the `commitButton.Text` update is not `updateCommitButtonState()`.

That said it'd probably best to split the buttons into two as suggested above because stuff like the `commitOrLogIn()` method are quite... icky.

Reply:
not a nitpick at all, thanks for pointing that out. The distinction between the two `updateState`  methods felt quite bad - luckily splitting the button in two removes the need for any cross-calling between the two.